### ISSUE TITLE
Same-origin requests are classified as cross-origin

### ISSUE THREAD

**Issue by petterdaae** [spring-projects/spring-framework#31277]
Hello 👋 

The [isCorsReuest](https://github.com/spring-projects/spring-framework/blob/d46c26d90329c3419ece77b3e11e3986e6184726/spring-web/src/main/java/org/springframework/web/cors/CorsUtils.java#L41) utility classifies our same-origin requests as cross-origin because we use the `Host` header for routing in our internal infrastructure. The problem is that *isCorsRequest* compares the `Host` with the `Origin` header to determine this, and these will never be the same in our case. This means that we have to specify the origins that our application is hosted on as allowed origins, and this quickly becomes a headache when you have to maintain this list across multiple micro-services that are hosted on multiple origins.

I am not aware of any specification that describes this behaviour. The [fetch standard](https://fetch.spec.whatwg.org/#http-cors-protocol) does not mention anything about comparing these headers to determine if it is a cross-origin request, and multiple sources claim that it is the browser (not the server) that should determine if a request should be rejected:

- [OWASP](https://owasp.org/www-project-web-security-testing-guide/v41/4-Web_Application_Security_Testing/11-Client_Side_Testing/07-Testing_Cross_Origin_Resource_Sharing) *... Based on the result of the OPTIONS request, the browser decides whether the request is allowed or not.*
- [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) *... allows a server to indicate any origins (domain, scheme, or port) other than its own from which a browser should permit loading resources.*
- [Wikipedia](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) *... the specification mandates that browsers "preflight" the request, soliciting supported methods from the server with an HTTP OPTIONS request method, and then, upon "approval" from the server, sending the actual request ...*

I am aware of the `ForwardedHeaderFilter`, but I think this is a somewhat fragile solution to the problem because it relies heavily on the proxies to set these headers correctly. Also, we would probably have to whitelist the content of these headers to mitigate host header injection attacks.

**Comment by sdeleuze** [comment]
> So, our same-origin requests are rejected by the server.

Could you please provide a reproducer (potentially a documented `http` or `curl` command line to use) to allow us to fully understand your use case?

**Comment by petterdaae** [comment]
My endpoint:
```kotlin
@RestController
class HealthController {

    @GetMapping("/health")
    fun health(): String = "Healthy\n"
}
```

My spring security config:
```kotlin
@Configuration
@EnableWebSecurity
class WebSecurityConfig {

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http.invoke {
            authorizeHttpRequests {
                authorize(anyRequest, permitAll)
            }
            cors { }
        }
        return http.build()
    }

    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val configuration = CorsConfiguration()
        configuration.allowedOrigins = listOf("http://example.com")
        configuration.allowedMethods = listOf("GET", "POST")
        val source = UrlBasedCorsConfigurationSource()
        source.registerCorsConfiguration("/**", configuration)
        return source
    }
}
```

When I run the below command, I expect my endpoint to reply normally, however, I get `Invalid CORS request` and 403 Forbidden.
```
curl http://localhost:8080/health -H"Origin: http://localhost:8080" -H"Host: internal.application.name.com"
```

(`http://example.com` is not relevant to mye question by the way, just wanted include a basic cors configuration)

*Is this enough, or do you want me to create git repository with the full source code of the project?*

**Comment by sdeleuze** [comment]
I would appreciate a repo or attached source code if possible.

**Comment by petterdaae** [comment]
https://github.com/petterdaae/spring-reproduce-cors

**Comment by petterdaae** [comment]
First, thank you for a thorough answer! Really appreciate it 🙌 

I can totally see the benefits of blocking simple requests and that `isCorsRequest` probably is the most reasonable implementation in this case. I think my main concern with this behaviour is that people working with spring-security get a false impression of what cors *really* is, but this is of course just my opinion and I can for sure see the security benefits that you get! 😅

Do you think a solution could be to implement something like `isSimpleRequest` and only block the request if it satisfies `isCorsRequest $$ isSimpleRequest`? In our case, we mainly use `application/json` in the `Content-Type` header which would not qualify as a simple request - and hence, there is no security benefit of blocking it. This would of course make things more complex, so I am not sure what I think myself.

**Comment by rstoyanchev** [comment]
@petterdaae are you looking to turn off CORS checks completely for your scenario? If so, you could plug in your own `CorsProcessor` to achieve that effect.

**Comment by rstoyanchev** [comment]
So some requests originate from browsers and others from within the firewall then? Is it feasible to narrow down with URL patterns in which case you can configure a more open policy for some, see [reference docs](https://docs.spring.io/spring-framework/reference/web/webmvc-cors.html#mvc-cors-global-java).

**Comment by petterdaae** [comment]
Oh, so it would for example be possible to have CORS on `/some-path/*` and skip the CORS handling entirely on all other requests?

**Comment by rstoyanchev** [comment]
Yes, you it's possible to set up CORS handling centrally based on URL patterns. This can be layered on top of more fine-grained config through annotations in controllers.

**Comment by sdeleuze** [comment]
The risk of lowering the security and introducing regressions seems too high to me, I prefer to decline this issue. Hopefully our feedback have helped. Thanks for your understanding.
